### PR TITLE
Some initial work for FlutterWeb apps

### PR DESCRIPTION
**Pull request by jwren**

### CODE REVIEW

**Comment by jwren** [comment]
@jacob314 Some additional comments and questions will be via email/ our chat.

File: src/io/flutter/run/SdkFields.java [hunk 1]
```diff
@@ -128,7 +129,13 @@ public GeneralCommandLine createFlutterSdkRunCommand(@NotNull Project project,
     if (buildFlavor != null) {
       args = ArrayUtil.append(args, "--flavor=" + buildFlavor);
     }
-    final FlutterCommand command = flutterSdk.flutterRun(root, main.getFile(), device, runMode, flutterLaunchMode, project, args);
+
+    final FlutterCommand command;
+    if(FlutterUtils.declaresFlutterWeb(root.getPubspec())) {
```
Comment (jacob314) [inline src/io/flutter/run/SdkFields.java]: nit:
```
final FlutterCommand command = FlutterUtils.declaresFlutterWeb(root.getPubspec() ?
  flutterSdk.flutterRunWeb(root, runMode, false) :
  flutterSdk.flutterRun(root, main.getFile(), device, runMode, flutterLaunchMode, project, args);
``` 
is a little simpler

File: src/io/flutter/run/SdkRunConfig.java [hunk 2]
```diff
@@ -142,8 +142,6 @@ public LaunchState getState(@NotNull Executor executor, @NotNull ExecutionEnviro
     final RunMode mode = RunMode.fromEnv(env);
     final Module module = ModuleUtilCore.findModuleForFile(mainFile.getFile(), env.getProject());
     final LaunchState.CreateAppCallback createAppCallback = (device) -> {
-      if (device == null) return null;
```
Comment (jacob314) [inline src/io/flutter/run/SdkRunConfig.java]: Does removing this check cause any problems for regular flutter apps?
An alternate that might be more robust is to create a mock ChromeBrowser Device to use for FlutterWeb.

File: src/io/flutter/run/daemon/FlutterApp.java [hunk 3]
```diff
@@ -68,7 +68,8 @@
   private final @NotNull Project myProject;
   private final @Nullable Module myModule;
   private final @NotNull RunMode myMode;
-  private final @NotNull FlutterDevice myDevice;
+  // TODO(github.com/flutter/flutter-intellij/issues/3293) myDevice is not-null for all run configurations except flutter web configurations
```
Comment (jacob314) [inline src/io/flutter/run/daemon/FlutterApp.java]: The bug listed doesn't cover the issue. e.g. whether Device should be null for FlutterWeb.

File: src/io/flutter/sdk/FlutterCommand.java [hunk 4]
```diff
@@ -48,13 +48,28 @@
   @NotNull
   private final List<String> args;
 
+  // TODO(github.com/flutter/flutter-intellij/issues/3293) This is a temporary "<pub-cache>/bin/webdev" that can be provided to test some
+  //  webdev directly.
+  @Nullable
+  private final String localWebDevExe = "webdev";
```
Comment (jacob314) [inline src/io/flutter/sdk/FlutterCommand.java]: can this be removed before checkin?

File: src/io/flutter/sdk/FlutterSdk.java [hunk 5]
```diff
@@ -269,6 +269,23 @@ else if (flutterLaunchMode == FlutterLaunchMode.RELEASE) {
...
 
+  public FlutterCommand flutterRunWeb(@NotNull PubRoot root, @NotNull RunMode mode, boolean hotReload) {
+    // flutter packages pub global run webdev serve [--debug] [--hot-reload]
+    final List<String> args = new ArrayList<>();
+    args.add("global");
+    args.add("run");
+    args.add("webdev");
+    args.add("daemon");
+    if (mode == RunMode.DEBUG) {
+      //args.add("--debug");
```
Comment (jacob314) [inline src/io/flutter/sdk/FlutterSdk.java]: do these need to be commented out? I imagine they might apply if/when we support debug mode.

File: src/io/flutter/server/vmService/DartVmServiceDebugProcess.java [hunk 6]
```diff
@@ -262,17 +263,21 @@ public void scheduleConnect() {
...
-      try {
-        vmService = VmService.connect(url);
-        vmOpenSourceLocationListener = VmOpenSourceLocationListener.connect(url);
-      }
-      catch (IOException | RuntimeException e) {
-        onConnectFailed("Failed to connect to the VM observatory service at: " + url + "\n"
-                        + e.toString() + "\n" +
-                        formatStackTraces(e));
-        return;
+      // TODO(github.com/flutter/flutter-intellij/issues/3293) The following check disables the WebSocket connection for all FlutterWeb
```
Comment (jacob314) [inline src/io/flutter/server/vmService/DartVmServiceDebugProcess.java]: please add more granularity to this issue. How does this test fail. 
If it is needed, check based on whether the app is a FlutterWebApp not based on the device being null.

File: src/io/flutter/run/SdkRunConfig.java [hunk 7]
```diff
@@ -142,6 +142,9 @@ public LaunchState getState(@NotNull Executor executor, @NotNull ExecutionEnviro
     final RunMode mode = RunMode.fromEnv(env);
     final Module module = ModuleUtilCore.findModuleForFile(mainFile.getFile(), env.getProject());
     final LaunchState.CreateAppCallback createAppCallback = (device) -> {
+      // Up until the FlutterWeb support, device was checked for null and returned.  The device can only be null if this is a FlutterWeb
```
Comment (jacob314) [inline src/io/flutter/run/SdkRunConfig.java]: nit: wrap comments at 80 chars.

File: src/io/flutter/run/daemon/FlutterApp.java [hunk 8]
```diff
@@ -597,6 +597,10 @@ public String deviceId() {
     return myDevice != null ? myDevice.deviceId() : "";
   }
 
+  public boolean isWebDev() {
+    return myDevice == null;
```
Comment (jacob314) [inline src/io/flutter/run/daemon/FlutterApp.java]: nit: add a TODO that this is a temp hack.

File: src/io/flutter/run/SdkRunConfig.java (hunk 7 above)
Comment (jwren) [reply src/io/flutter/run/SdkRunConfig.java]: wrapped

File: src/io/flutter/run/daemon/FlutterApp.java (hunk 8 above)
Comment (jwren) [reply src/io/flutter/run/daemon/FlutterApp.java]: TODO added

**Pull request merged**